src/tile.cpp

```diff
@@ -1405,30 +1393,30 @@ void Tile::postRemoveNotification(Thing* thing, const Cylinder* newParent, int32
 		onUpdateTile(spectators);
 	}
 
-	for (Creature* spectator : spectators) {
-		assert(dynamic_cast<Player*>(spectator) != nullptr);
-		static_cast<Player*>(spectator)->postRemoveNotification(thing, newParent, index, LINK_NEAR);
+	for (auto spectator : spectators) {
+		assert(spectator->getPlayer() != nullptr);
+		std::static_pointer_cast<Player>(spectator)->postRemoveNotification(thing, newParent, index, LINK_NEAR);
 	}
 
 	// calling movement scripts
-	Creature* creature = thing->getCreature();
+	auto creature = thing->getCreature();
 	if (creature) {
-		g_moveEvents->onCreatureMove(creature, this, MOVE_EVENT_STEP_OUT);
+		g_moveEvents->onCreatureMove(creature, getTile(), MOVE_EVENT_STEP_OUT);
 	} else {
-		Item* item = thing->getItem();
+		auto item = thing->getItem();
 		if (item) {
-			g_moveEvents->onItemMove(item, this, false);
+			g_moveEvents->onItemMove(item, getTile(), false);
 		}
 	}
 }
 
-void Tile::internalAddThing(Thing* thing) { internalAddThing(0, thing); }
+void Tile::internalAddThing(std::shared_ptr<Thing> thing) { internalAddThing(0, std::move(thing)); }
 
-void Tile::internalAddThing(uint32_t, Thing* thing)
+void Tile::internalAddThing(uint32_t, std::shared_ptr<Thing> thing)
 {
-	thing->setParent(this);
+	thing->setParent(getTile());
 
-	Creature* creature = thing->getCreature();
+	auto creature = thing->getCreature();
```

Comment:
Could this be `const auto`
```suggestion
	const auto creature = thing->getCreature();
```